<a href="https://colab.research.google.com/github/gitty-scientiest/IBM-module-2-/blob/main/And_hyd_pynb.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

# Reading and understand The data

In [2]:
df = pd.read_csv("/content/DirectRouting.csv")

In [3]:
df #checking the dataframe

Start time                      End time  \
0                              NaN                           NaN   
1                              NaN                           NaN   
2                              NaN                           NaN   
3                              NaN                           NaN   
4                              NaN                           NaN   
...                            ...                           ...   
8118  2024-02-06T23:13:11.2550000Z                           NaN   
8119  2024-02-06T23:24:13.7540000Z                           NaN   
8120  2024-02-06T23:24:16.6210000Z                           NaN   
8121  2024-02-06T23:27:49.9950000Z  2024-02-06T23:28:39.9950000Z   
8122  2024-02-06T23:28:53.8760000Z                           NaN   

                Duration (seconds) Success  
0                                0      No  
1                                0      No  
2                                0      No  
3                                0      No  
4                                0      No  
...                            ...     ...  
8118  2024-02-06T23:13:55.3090000Z      11  
8119  2024-02-06T23:25:00.8370000Z      11  
8120  2024-02-06T23:25:00.5860000Z      11  
8121                            50     Yes  
8122  2024-02-06T23:30:07.9300000Z      49  

[8123 rows x 4 columns]

In [4]:
print('shape of the dataframe:')
df.shape

shape of the dataframe:


(8123, 4)

In [5]:
df.info()#knowing the dataframe

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8123 entries, 0 to 8122
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Start time          7378 non-null   object
 1   End time            2845 non-null   object
 2   Duration (seconds)  5925 non-null   object
 3   Success             8120 non-null   object
dtypes: object(4)
memory usage: 254.0+ KB


Pre-prossesing the data and filtering out  the Unsuccessful cells according to commands .

In [6]:
# creating a new data frame with only successful data cells
df1 = df[df['Success']=='Yes'] #we can also use here if loop
#if success is yes than print ( but it will time taking )

In [7]:
df1 # checking the new dataframe

Start time                      End time  \
11    2024-02-05T04:53:47.5080000Z  2024-02-05T04:54:34.5080000Z   
13    2024-02-05T05:09:16.7810000Z  2024-02-05T05:09:19.7810000Z   
14    2024-02-05T05:09:19.8750000Z  2024-02-05T05:09:35.8750000Z   
15    2024-02-05T05:09:19.7340000Z  2024-02-05T05:09:35.7340000Z   
16    2024-02-05T05:18:05.9100000Z  2024-02-05T05:18:32.9100000Z   
...                            ...                           ...   
8064  2024-02-06T21:04:38.4640000Z  2024-02-06T21:06:29.4640000Z   
8099  2024-02-06T22:20:19.9440000Z  2024-02-06T22:21:32.9440000Z   
8108  2024-02-06T22:51:51.9210000Z  2024-02-06T23:06:39.9210000Z   
8116  2024-02-06T23:11:45.2900000Z  2024-02-06T23:19:59.2900000Z   
8121  2024-02-06T23:27:49.9950000Z  2024-02-06T23:28:39.9950000Z   

     Duration (seconds) Success  
11                   47     Yes  
13                    3     Yes  
14                   16     Yes  
15                   16     Yes  
16                   27     Yes  
...                 ...     ...  
8064                111     Yes  
8099                 73     Yes  
8108                888     Yes  
8116                494     Yes  
8121                 50     Yes  

[639 rows x 4 columns]

In [8]:
print('new dataframe shape:')
df1.shape

new dataframe shape:


(639, 4)

In [9]:
df1.info() #understanding new dataframe

<class 'pandas.core.frame.DataFrame'>
Index: 639 entries, 11 to 8121
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Start time          639 non-null    object
 1   End time            639 non-null    object
 2   Duration (seconds)  639 non-null    object
 3   Success             639 non-null    object
dtypes: object(4)
memory usage: 41.1+ KB


Hence we have to work only on 639 cells out of 8123 as only these cells are successful .

#Fixing the data columns in the correct datatype.

In [10]:
from datetime import datetime

In [11]:
#converting to datetime
df1['Start time'] = pd.to_datetime(df1['Start time'])
df1['End time'] = pd.to_datetime(df1['End time'])

<ipython-input-11-a9af4223b958>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1['Start time'] = pd.to_datetime(df1['Start time'])
<ipython-input-11-a9af4223b958>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1['End time'] = pd.to_datetime(df1['End time'])


In [12]:
df1['Start time'].dtype # checking the datatype

datetime64[ns, UTC]

In [13]:
df1['End time'].dtypes

datetime64[ns, UTC]

#Implementing a function to calculate the maximum number of parallel calls:

In [14]:
from collections import defaultdict
from operator import itemgetter
from itertools import groupby

In [15]:
from collections import defaultdict
def max_pc(df):  # calculate the max no of pc
    timestamps = defaultdict(int)

    # Increment and decrement call counts based on start and end times
    for _, row in df.iterrows():
        timestamps[convert_to_minutes(row['Start time'])] += 1
        timestamps[convert_to_minutes(row['End time'])] -= 1

    max_pc = 0
    current_pc = 0

    # Iterate through timestamps to calculate the maximum parallel calls
    for time in sorted(timestamps):
        current_pc += timestamps[time]
        max_pc = max(max_pc, current_pc)

    return max_pc

In [16]:
def group_calls_by_minute(df):  # group calls by minutes
    calls_by_minute = defaultdict(list)
    for _, row in df1.iterrows():  # Corrected indentation
        for t in range(convert_to_minutes(row['Start time']), convert_to_minutes(row['End time'])):
            calls_by_minute[t].append((row['Start time'], row['End time']))
    return calls_by_minute

In [17]:
def convert_to_minutes(timestamp): #Convert timestamp to minutes.



  return int(timestamp.timestamp() / 60)

# Calculate the maximum number of parallel calls
max_parallel_calls = max_pc(df1)

# Print the result
print(f'The maximum number of parallel calls needed is: {max_parallel_calls}')

The maximum number of parallel calls needed is: 5


In [18]:
def convert_to_minutes(timestamp):
    if isinstance(timestamp, float):
        return int(timestamp / 60)
    else:
        return int(timestamp.timestamp() / 60)

# Calculate the maximum number of parallel calls
max_parallel_calls = max_pc(df1)

# Print the result
print(f'The maximum number of parallel calls needed is: {max_parallel_calls}')

The maximum number of parallel calls needed is: 5
